m3.2_linear regression

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [2]:
myData = pd.read_csv('3.2_mycar.csv')

In [3]:
X = myData.iloc[:,:-1].values
Y = myData.iloc[:,1].values

In [4]:
from sklearn.model_selection import train_test_split

получаем независимые и зависимые переменные из обучающей и тестовой выборки. Размер тестовой выборки задаем 0,3.

In [6]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size= 0.3)

обучим модель:

In [8]:
from sklearn.linear_model import LinearRegression
myModel = LinearRegression() #Обозначаем, что наша модель - линейная регрессия
myModel.fit(X_train,Y_train) #обучаем модель на обучающих данных

LinearRegression()

In [9]:
y_pred = myModel.predict(X_test)
y_pred

array([23.35, 68.2 , 38.3 , 65.21, 20.36, 35.31, 35.31, 29.33, 47.27,
       26.34, 14.38, 53.25, 47.27,  5.41, 38.3 ])

3.3 tasks - below

3.3.1 

In [23]:
A = [1,3,2,5] #prognose value
B = [2,3,-1,4] #true value
n = 4
ab = 0
for i in range(0,4):
    ab += (A[i]-B[i])**2
    print((A[i]-B[i])**2)

1
0
9
1


In [24]:
MSE = (1/n)*ab
MSE #correct

2.75

3.4 tasks - below

Для каких данных лучше использовать OLS?
- Если в обучающей выборке мало признаков (верно)

(альфа - это темп обучения)

#### 3.4.2 Найдите следующий шаг градиентного спуска. Текущая модель: y=2x; Темп обучения (learning rate): 1/6; Ответ будет выглядеть как y=kx+b, где k и b вы считаете самостоятельно. 

y = betta1*x + betta2
y = 2x + 0 --> betta_k_minus_one = np.array([[2],[0]])

In [173]:
x = np.array([1,2]) #матрица признаков
x = np.array([[1,1], [2,1]]) #матрица признаков - добавим к  столбец единицу, чтобы учесть слагаемое betta2
x1 = (1,1)
y = (2,5)
y = np.array([2,5])
alfa = 1/6
#betta = (x.transpose()*x)^(-1) *x.transpose()*y

In [181]:
summa = 0
betta_k_minus_one = np.array([2,0])
for i in range(0,2):
    summa += x[i].transpose() * (sum(x[i]*betta_k_minus_one) - y[i])

In [182]:
summa

array([-2, -1])

In [179]:
betta_k_minus_one + ([1/3, 1/6])

array([2.33333333, 0.16666667])

In [185]:
n = 2
betta_k = betta_k_minus_one - 2*alfa*(1/n)*summa
betta_k #correct

array([2.33333333, 0.16666667])

then betta1 = 7/3 = 2.33333333, betta2 = 1/6 = 0.16666667
- then y = 7/3*x +1/6

#### 3.4.5 Чему равен коэффициент детерминации на этих данных? Укажите число с точностью до сотых:

In [216]:
fi_list = [1,3,2,5] #prognosed value
yi_list = [2,3,-1,4] #true value

In [217]:
#v1
###
y_mean = sum(yi_list)/float(len(yi_list))
ss_tot = sum((yi-y_mean)**2 for yi in yi_list)
ss_err = sum((yi-fi)**2 for yi,fi in zip(yi_list,fi_list))
r2 = 1 - (ss_err/ss_tot)
###
r2 #correct

0.2142857142857143

In [212]:
#Коэффициент детерминации
SS_res = ss_err
SS_tot = ss_tot

In [213]:
R_square = 1 - SS_res/SS_tot #доля дисперсии зависимой переменной
R_square

0.2142857142857143